external/fv3fit/tests/keras/test_halos.py

```diff
@@ -0,0 +1,159 @@
+    check_result(result=result, n_halo=n_halo, nx=nx, ny=ny, nz=nz)
+    assert result.dims.keys() == ds.dims.keys()
+
+
+def check_result(result: xr.Dataset, n_halo: int, nx: int, ny: int, nz: int):
+    for name, da in result.data_vars.items():
+        assert da.sizes["x"] == nx + 2 * n_halo, name
+        assert da.sizes["y"] == ny + 2 * n_halo, name
+        assert da.sizes.get("z", nz) == nz, name
+        # only corners should still be zero
+        compute_data = da.isel(
+            x=range(n_halo, n_halo + nx), y=range(n_halo, n_halo + ny)
+        )
+        # compute data was never zero, shouldn't be now
+        assert np.sum(compute_data == 0) == 0, name
+        # x and y edge halos should be 1 now, but corners are still zero
+        data_with_x_halos = da.isel(
+            x=range(0, nx + 2 * n_halo), y=range(n_halo, n_halo + ny)
+        )
+        assert np.sum(data_with_x_halos == 0) == 0, name
+        data_with_y_halos = da.isel(
+            x=range(n_halo, n_halo + nx), y=range(0, ny + 2 * n_halo)
+        )
+        assert np.sum(data_with_y_halos == 0) == 0, name
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_tile, n_halo", [pytest.param(8, 8, 12, 1, 3, id="typical")]
+)
+def test_append_halos_raises_with_one_tile(
+    nx: int, ny: int, nz: int, n_tile: int, n_halo: int
+):
+    ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile)
+    with pytest.raises(ValueError):
+        append_halos(ds, n_halo=n_halo)
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_tile", [pytest.param(8, 8, 12, 6, id="typical")]
+)
+def test_append_halos_no_halos_is_unchanged(nx: int, ny: int, nz: int, n_tile: int):
+    ds = get_dataset(nx=nx, ny=ny, nz=nz, n_tile=n_tile)
+    result: xr.Dataset = append_halos(ds, n_halo=0)
+    xr.testing.assert_identical(result, ds)
+
+
+class ChainedDummyComm(fv3gfs.util.testing.DummyComm):
+    """
+    Dummy comm that calls a callback just before recv, allowing us to call all sends
+    before recvs.
+
+    Can be used to mock MPI behavior in a function that has a single send-recv pair.
+    """
+
+    def __init__(self, rank, total_ranks, buffer_dict):
+        super().__init__(rank=rank, total_ranks=total_ranks, buffer_dict=buffer_dict)
+        self._callback = lambda: None
+
+    def set_callback(self, callback):
+        self._callback = callback
+
+    def Recv(self, *args, **kwargs):
+        self._callback()
+        self._callback = lambda: None
+        super().Recv(*args, **kwargs)
+
+
+def _get_callback(output_datasets, i: int, rank_dataset, n_halo, comm):
+    def callback():
+        output_datasets[i] = _append_halos_using_mpi(
+            ds=rank_dataset, n_halo=n_halo, comm=comm
+        )
+
+    return callback
+
+
+@pytest.mark.parametrize(
+    "nx, ny, nz, n_halo",
+    [pytest.param(8, 8, 12, 3, id="typical"), pytest.param(8, 8, 12, 1, id="one_halo")],
+)
+def test_with_and_without_mpi_give_same_result(nx: int, ny: int, nz: int, n_halo: int):
```

Comment:
Clever idea to test this way. Just to make sure I understand, the test adds a daisy-chain blocking sequence to the mpi-like case to simulate real MPI async behavior despite the fact that it's actually using a dummy com? And `Recv` is triggered by the halo update? Because outside of the comm the mpi- and non-mpi halo updates aren't really different so I can see that this is kind of hard to test.

Reply:
Yesss, exactly. Each rank will run until it needs to receive data then tell the next one to run completely before it continues. It doesn't work if there's more than 1 send-receive pairs. In the long run we'll hopefully have a solution that uses threading and a shared dictionary, where the thread can wait for other ranks/threads.